### ISSUE TITLE
3D platformer demo crashes with GLES2 on Android/Mali

### ISSUE THREAD

**Issue by akien-mga** [godotengine/godot#26972]
**Godot version:**
53131635b

**OS/device including version:**
[Samsung Galaxy S3](http://opengles.gpuinfo.org/displayreport.php?id=3407), Mali-400 MP, Android 7.1.2
[Samsung Galaxy Tab A (2016)](http://opengles.gpuinfo.org/displayreport.php?id=3450), Mali-T830, Android 8.1.0 (from @ProbDenis)

**Issue description:**
When running the [3D platformer demo](https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer) on the above Android devices with GLES2, it crashes with the following backtrace (extracted with a `target=debug` build):
```
03-12 12:19:12.653  6405  6423 E godot   : **ERROR**: R float texture not supported, converting to RGB8.
03-12 12:19:12.653  6405  6423 E godot   :    At: drivers/gles2/rasterizer_storage_gles2.cpp:162:_get_gl_image_and_format() - R float texture not supported, converting to RGB8.
03-12 12:19:12.653  6405  6423 E godot   : **ERROR**: R float texture not supported, converting to RGB8.
03-12 12:19:12.653  6405  6423 E godot   :    At: drivers/gles2/rasterizer_storage_gles2.cpp:162:_get_gl_image_and_format() - R float texture not supported, converting to RGB8.
03-12 12:19:12.661  6405  6423 E godot   : **ERROR**: R float texture not supported, converting to RGB8.
03-12 12:19:12.661  6405  6423 E godot   :    At: drivers/gles2/rasterizer_storage_gles2.cpp:162:_get_gl_image_and_format() - R float texture not supported, converting to RGB8.
03-12 12:19:12.661  6405  6423 E godot   : **ERROR**: R float texture not supported, converting to RGB8.
03-12 12:19:12.661  6405  6423 E godot   :    At: drivers/gles2/rasterizer_storage_gles2.cpp:162:_get_gl_image_and_format() - R float texture not supported, converting to RGB8.
03-12 12:19:13.096  6405  6423 E godot   : **WARNING**: The method set_friction has been deprecated and will be removed in the future, use physics material instead.
03-12 12:19:13.096  6405  6423 E godot   :    At: scene/3d/physics_body.cpp:640:set_friction() - This method has been deprecated and will be removed in the future
03-12 12:19:13.096  6405  6423 E godot   : **WARNING**: The method set_bounce has been deprecated and will be removed in the future, use physics material instead.
03-12 12:19:13.096  6405  6423 E godot   :    At: scene/3d/physics_body.cpp:669:set_bounce() - This method has been deprecated and will be removed in the future
03-12 12:19:13.227  6405  6423 E godot   : **WARNING**: AnimationTreePlayer has been deprecated. Use AnimationTree instead.
03-12 12:19:13.227  6405  6423 E godot   :    At: scene/animation/animation_tree_player.cpp:405:_notification() - This method has been deprecated and will be removed in the future
03-12 12:19:14.133  6405  6423 E godot   : **ERROR**: Condition ' !p_axis.is_normalized() ' is true.
03-12 12:19:14.133  6405  6423 E godot   :    At: core/math/basis.cpp:799:set_axis_angle() - Condition ' !p_axis.is_normalized() ' is true.
03-12 12:19:14.133  6405  6423 E godot   : **ERROR**: Condition ' p_eye == p_target ' is true.
03-12 12:19:14.133  6405  6423 E godot   :    At: core/math/transform.cpp:88:set_look_at() - Condition ' p_eye == p_target ' is true.
03-12 12:19:14.134  6405  6423 E godot   : **ERROR**: Condition ' !p_axis.is_normalized() ' is true.
03-12 12:19:14.134  6405  6423 E godot   :    At: core/math/basis.cpp:799:set_axis_angle() - Condition ' !p_axis.is_normalized() ' is true.
03-12 12:19:14.134  6405  6423 E godot   : **ERROR**: Condition ' p_eye == p_target ' is true.
03-12 12:19:14.134  6405  6423 E godot   :    At: core/math/transform.cpp:88:set_look_at() - Condition ' p_eye == p_target ' is true.
03-12 12:19:14.134  6405  6423 E godot   : **ERROR**: Condition ' !p_axis.is_normalized() ' is true.
03-12 12:19:14.134  6405  6423 E godot   :    At: core/math/basis.cpp:799:set_axis_angle() - Condition ' !p_axis.is_normalized() ' is true.
03-12 12:19:14.134  6405  6423 E godot   : **ERROR**: Condition ' p_eye == p_target ' is true.
03-12 12:19:14.134  6405  6423 E godot   :    At: core/math/transform.cpp:88:set_look_at() - Condition ' p_eye == p_target ' is true.
03-12 12:19:14.134  6405  6423 E godot   : **ERROR**: Condition ' !p_axis.is_normalized() ' is true.
03-12 12:19:14.135  6405  6423 E godot   :    At: core/math/basis.cpp:799:set_axis_angle() - Condition ' !p_axis.is_normalized() ' is true.
03-12 12:19:14.135  6405  6423 E godot   : **ERROR**: Condition ' p_eye == p_target ' is true.
03-12 12:19:14.135  6405  6423 E godot   :    At: core/math/transform.cpp:88:set_look_at() - Condition ' p_eye == p_target ' is true.
03-12 12:19:14.716  6405  6423 F libc    : Fatal signal 11 (SIGSEGV), code 1, fault addr 0xc in tid 6423 (GLThread 311)
03-12 12:19:14.719  1946  1946 W         : debuggerd: handling request: pid=6405 uid=10149 gid=10149 tid=6423
03-12 12:19:14.750  6443  6443 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
03-12 12:19:14.750  6443  6443 F DEBUG   : LineageOS Version: '14.1-20181117-NIGHTLY-i9300'
03-12 12:19:14.750  6443  6443 F DEBUG   : Build fingerprint: 'samsung/m0xx/m0:4.3/JSS15J/I9300XXUGMJ9:user/release-keys'
03-12 12:19:14.750  6443  6443 F DEBUG   : Revision: '0'
03-12 12:19:14.750  6443  6443 F DEBUG   : ABI: 'arm'
03-12 12:19:14.751  6443  6443 F DEBUG   : pid: 6405, tid: 6423, name: GLThread 311  >>> org.godotengine.platformer3d <<<
03-12 12:19:14.751  6443  6443 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0xc
03-12 12:19:14.751  6443  6443 F DEBUG   :     r0 0000000c  r1 00000000  r2 00000018  r3 00000000
03-12 12:19:14.751  6443  6443 F DEBUG   :     r4 42cef234  r5 00000007  r6 00000000  r7 00000000
03-12 12:19:14.751  6443  6443 F DEBUG   :     r8 518f3238  r9 518f2f00  sl 3457b9b1  fp 518f2d38
03-12 12:19:14.751  6443  6443 F DEBUG   :     ip 00fc26a1  sp 518f2928  lr 4fec33a4  pc 4fec3408  cpsr 600f0010
03-12 12:19:15.118  6443  6443 F DEBUG   : 
03-12 12:19:15.118  6443  6443 F DEBUG   : backtrace:
03-12 12:19:15.118  6443  6443 F DEBUG   :     #00 pc 00941408  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN20RasterizerSceneGLES215_setup_geometryEPNS_10RenderList7ElementEPN22RasterizerStorageGLES28SkeletonE+1112)
03-12 12:19:15.118  6443  6443 F DEBUG   :     #01 pc 0093fb9c  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN20RasterizerSceneGLES219_render_render_listEPPNS_10RenderList7ElementEiRK9TransformRK12CameraMatrix3RIDPNS_11EnvironmentEjffbbb+4092)
03-12 12:19:15.118  6443  6443 F DEBUG   :     #02 pc 00947d80  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN20RasterizerSceneGLES213render_shadowE3RIDS0_iPPN15RasterizerScene12InstanceBaseEi+4084)
03-12 12:19:15.118  6443  6443 F DEBUG   :     #03 pc 017517c4  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN17VisualServerScene29_light_instance_update_shadowEPNS_8InstanceE9TransformRK12CameraMatrixb3RIDPNS_8ScenarioE+5636)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #04 pc 0174e7d8  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN17VisualServerScene14_prepare_sceneE9TransformRK12CameraMatrixb3RIDjS4_S4_S4_+4048)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #05 pc 0174d558  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN17VisualServerScene13render_cameraE3RIDS0_7Vector2S0_+744)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #06 pc 01770d58  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN20VisualServerViewport14_draw_viewportEPNS_8ViewportEN13ARVRInterface4EyesE+984)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #07 pc 01772a50  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN20VisualServerViewport14draw_viewportsEv+1644)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #08 pc 017361a0  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN18VisualServerRaster4drawEbd+380)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #09 pc 0177af10  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN18VisualServerWrapMT4drawEbd+240)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #10 pc 000f5810  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN4Main9iterationEv+1560)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #11 pc 000992a8  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (_ZN10OS_Android17main_loop_iterateEv+52)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #12 pc 000ad31c  /data/app/org.godotengine.platformer3d-1/lib/arm/libgodot_android.so (Java_org_godotengine_godot_GodotLib_step+280)
03-12 12:19:15.119  6443  6443 F DEBUG   :     #13 pc 001157cd  /data/app/org.godotengine.platformer3d-1/oat/arm/base.odex (offset 0x109000)
```

Note: the "R float texture" errors are expected and due to the use of `CurveTexture` in a particle effect. They used to cause an earlier crash which was fixed by #26943, so this new crash is the next step to solve.

This new crash seems related with rendering shadows and/or handling skeletons.
It would be useful to try simpler 3D demos on such devices (or test removing features from the Platformer 3D until it works) to identify what component causes the crash exactly.

**Steps to reproduce:**
- Add Android preset to Platformer 3D, enable ETC import and switch to GLES2
- Export to one of the above phones
- Note: I got things to work fine on a higher end phone (Xiaomi Pocophone F1), though I had to change the physics backend to GodotPhysics, as Bullet would crash. I'll soon do that directly in the demo

**Minimal reproduction project:**
https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer

**Comment by ProbDenis** [comment]
I'll try to test different 3D scenes and see if I can narrow it down some more.

**Comment by ProbDenis** [comment]
It's definitely the skeletons. Here's a tiny example project, which has two scenes:
"Skeleton.tscn" has an imported cube (dae) with a skeleton.
"NoSkeleton.tscn" has an imported cube without a skeleton.
[Android_GLES2_Skeletons.zip](https://github.com/godotengine/godot/files/3086156/Android_GLES2_Skeletons.zip)

Now, if Skeleton.tscn is the main scene, the project crashes on my tablet on GLES2. But if NoSkeleton.tscn is the main scene, it doesn't crash. The skeleton (with an animation, which isn't played) is the only difference between the two scenes.

**Comment by akien-mga** [comment]
This is likely the same issue as #28298. Not closing as duplicate just yet, but we'll see if fixing one fixes the other.

**Comment by lawnjelly** [comment]
My [PR fixed version](https://github.com/godotengine/godot/pull/29751) will run the platform demo without crashing when I converted it to ES2, and added ETC textures. However the buttons show but the screen is white on my Nexus 7, there is probably a little tweaking it needs to show up well in ES2, but it looks promising (it would crash on mine too without the PR), probably just some lighting or world environment needs adjusting.

**Comment by lawnjelly** [comment]
![platformer](https://user-images.githubusercontent.com/21999379/59452312-58f5df80-8e05-11e9-9590-e695a357399b.jpg)
This is a screenshot of it working in GLES2 on my Nexus 7 2012. Once the skinning is working with the PR the issue is that you can't just change the renderer from GLES3 to 2 and expect it to work straight off, some shaders don't work in GLES2 etc. I tried switching off a few things and got it starting to work ok.

Really for demos you need a separate section for demos for GLES3 and another for GLES2, or some way of sharing some of the assets between them if you wanted to be more efficient, but that is probably overkill.

**Comment by Zireael07** [comment]
GLES 2 is not meant to be a fallback for GLES 3, especially shaders won't work cleanly.

**Comment by akien-mga** [comment]
Fixed by #29751.
